docs/quick_start.md

```diff
@@ -0,0 +1,316 @@
+   `spider start --worker --db <db_url>`. Starting a worker that can run specific tasks needs to
+   link to libraries. We'll cover this later.
+
+## Start a client
+
+Client first creates a Spider client driver and connects it to the database. Spider automatically
+cleans up the resource in driver's destructor.
+
+```c++
+#include <spider/Spider.hpp>
+
+auto main(int argc, char **argv) -> int {
+    spider::Driver driver{"db_url"};
+}
+```
+
+## Create a task
+
+In Spider, a task is a non-member function that takes the first argument a `spider::Context` object.
+It can then take any number of arguments of POD type or `spider::Data` covered
+in [later section](#data-on-external-storage).
+
+Task can return any POD type. If a task needs to return more than one result, uses `std::tuple` and
+makes sure all elements of `std::tuple` are POD or `spider::Data`.
+
+The `Context` object represents the context of a running task. It provides methods to get the task
+metadata information like task id. It also supports the creating task inside a task. We will cover
+this later.
+
+```c++
+// Task that sums to integers
+auto sum(spider::Context &context, int x, int y) -> int {
+    return x + y;
+}
+
+// Task that sorts two integers in non-acesending order
+auto sort(spider::Context &context, int x, int y) -> std::tuple<int, int> {
+    if (x >= y) {
+        return { x, y };
+    }
+    return { y, x };
+}
+```
+
+## Run a task
+
+Spider enables user to run a task on the cluster. First register the functions statically so it is
+known by Spider. Simply call `Driver::run` and provide the arguments of the task. `Driver::run`
+returns a `spider::Future` object, which represents the result that will be available in the future.
+You can call `Future::ready` to check if the value in future is available yet. You can use
+`Future::get` to block and get the value once it is available.
+
+```c++
+spider::register_task(sum);
+spider::register_task(sort);
+
+auto main(int argc, char **argv) -> int {
+    // driver initialization skipped
+    spider::Future<int> sum_future = driver.run(sum, 2);
+    assert(4 == sum_future.get());
+
+    spider::Future<std::tuple<int, int>> sort_future = driver.run(4, 3);
+    assert(std::tuple{3, 4} == sort_future.get());
+}
+```
+
+## Group tasks together
+
+In real world, running a single task is too simple to be useful. Spider lets you bind outputs of
+tasks as inputs of another task, similar to `std::bind`. The first argument of `spider::bind` is the
+child task. The later arguments are either a `spider::Task` or a `spider::TaskGraph`, whose entire
+outputs are used as part of the inputs to the child task, or a POD or
+`spider::Data` that is directly used as input. Spider requires that the types of `Task` or
+`TaskGraph` outputs or POD type or `spider::Data` matches the input types of child task.
+
+Binding the tasks together forms a dependencies among tasks, which is represented by
```

Comment:
1. How do you plan to communicate (serialize) the task calls (name + args) from the client to the server?
2. Let's say a task takes two inputs. Does this interface support using a constant for one input and a task for the other?
   1. If so, do you support a task graph that looks like this?
      ```mermaid
      flowchart TD
          leaf["foo(int, int) -> int"]
          parent["bar(int, int) -> int"]
          3 --> leaf
          4 --> parent
          5 --> parent
          parent --> leaf
      ```
   2. If so, that means any arguments the user passes into `run` get passed to the inputs in the task graph with a kind of DFS ordering. Is that true?

Reply:
1. Here comes the use of `spider::register_task`, which records the mapping between function name and the function pointer. When the user calls `run` with the function pointer, the library actually sends the function name to the db. Worker also gets the name of the function as part of the task metadata. Since the worker links to the library with `register_task` call, it also has the mapping and can know which function to call.
As for arguments, their types are stored in db (serialized as string right now), and values are serialized into string.
2. Yes.
  i. Yes.
 ii. No. `spider::bind` needs to bind all inputs of the child task. Thus, the value 3 must be the an argument in bind. 4 and 5 can be passed in `run`. However, it is possible that there are multiple first layer tasks, i.e. tasks with no parents. The input of task graph is the product of all first layer tasks' output.
 It is possible to support passing 3 as an argument in `run` for the above example. We can have a special placeholder type for `bind`. In such case, DFS is an intuitive order.

Reply:
1. Can you explain how:
   1. `register_task` maps a function pointer into a function name?
   2. `run` turns a function pointer into a function name?
2. Can you explain this point with an example?
   > However, it is possible that there are multiple first layer tasks, i.e. tasks with no parents. The input of task graph is the product of all first layer tasks' output.

Reply:
1.  For client and worker, when the library is loaded, it will create a mapping between function name and function pointer.
  i. I am thinking of turning `register_task` into a macro to get the function name at compile time. It then stores the mapping between the function name to the function pointer.
  ii. `run` gets the function name from the mapping stored before.
2. In the example below, both `bar` and `baz` has no parent, and task graph input is their inputs put together, i.e. 1, 2, 3 and 4.
```mermaid
graph TD
    1[1]
    2[2]
    3[3]
    4[4]
    foo["foo(int, int) -> int"]
    bar["bar(int, int) -> int"]
    baz["baz(int, int) -> int"]
    1 --> bar
    2 --> bar
    3 --> baz
    4 --> baz
    bar --> foo
    baz --> foo
```

Reply:
1. Makes sense.
2. Gotcha. So to clarify my understanding:
   * in `run`, users can only pass arguments to root tasks (tasks with no parents).
   * in `run`, when the user passes a list of arguments `[1, 2, 3, 4]`, conceptually:
     * the framework will iterate over the root tasks in order;
     * for each task argument, the framework will pop one argument from the list.

Reply:
Yes and yes.